gkr_iop/src/precompiles/lookup_keccakf.rs

```diff
@@ -670,99 +690,119 @@ impl<E: ExtensionField> ProtocolBuilder<E> for KeccakLayout<E> {
-
-        assert_eq!(global_and_lookup, 3 * AND_LOOKUPS);
-        assert_eq!(global_xor_lookup, 3 * AND_LOOKUPS + 3 * XOR_LOOKUPS);
-        assert_eq!(global_range_lookup, LOOKUP_FELTS_PER_ROUND);
-
         let keccak_input8: ArrayView<(WitIn, EvalExpression<E>), Ix3> =
             ArrayView::from_shape((5, 5, 8), keccak_input8).unwrap();
         let keccak_input32 = keccak_input32.to_vec();
         let mut keccak_input32_iter = keccak_input32.iter().cloned();
 
-        // process keccak input
+        let keccak_output32 = keccak_output32.to_vec();
+        let keccak_output8: ArrayView<(WitIn, EvalExpression<E>), Ix3> =
+            ArrayView::from_shape((5, 5, 8), keccak_output8).unwrap();
+        let mut keccak_output32_iter = keccak_output32.iter().cloned();
+
+        // process keccak output
         for x in 0..5 {
             for y in 0..5 {
                 for k in 0..2 {
                     // create an expression combining 4 elements of state8 into a single 32-bit felt
                     let expr = expansion_expr::<E, 32>(
-                        keccak_input8
+                        &keccak_output8
                             .slice(s![x, y, 4 * k..4 * (k + 1)])
                             .iter()
                             .map(|e| (8, e.0.expr()))
-                            .collect_vec()
-                            .as_slice(),
+                            .collect_vec(),
+                    );
+                    system.add_non_zero_constraint(
+                        expr,
+                        keccak_output32_iter.next().unwrap().clone(),
+                        format!("build 32-bit output: {x}, {y}, {k}"),
                     );
-                    expressions.push(expr);
-                    evals.push(keccak_input32_iter.next().unwrap().clone());
-                    expr_names.push(format!("build 32-bit input: {x}, {y}, {k}"));
                 }
             }
         }
 
-        let keccak_output32 = keccak_output32.to_vec();
-        let keccak_output8: ArrayView<(WitIn, EvalExpression<E>), Ix3> =
-            ArrayView::from_shape((5, 5, 8), keccak_output8).unwrap();
-        let mut keccak_output32_iter = keccak_output32.iter().cloned();
-
-        // process keccak output
+        // process keccak input
         for x in 0..5 {
             for y in 0..5 {
                 for k in 0..2 {
                     // create an expression combining 4 elements of state8 into a single 32-bit felt
                     let expr = expansion_expr::<E, 32>(
-                        &keccak_output8
+                        keccak_input8
                             .slice(s![x, y, 4 * k..4 * (k + 1)])
                             .iter()
                             .map(|e| (8, e.0.expr()))
-                            .collect_vec(),
+                            .collect_vec()
+                            .as_slice(),
+                    );
+                    system.add_non_zero_constraint(
```

Comment:
all the previous hacky operation:
- expressions.push
- expr_names.push
- evals.push
are condense into just one call `add_non_zero_constraint` for error pruning